common/ringpop/ringpop.go

```diff
@@ -34,131 +35,186 @@ import (
-func (factory *RingpopFactory) GetMembershipMonitor() (membership.Monitor, error) {
-	factory.Lock()
-	defer factory.Unlock()
+// ValidateRingpopConfig validates that ringpop config is parseable and valid
+func ValidateRingpopConfig(rpConfig *config.Membership) error {
+	if rpConfig.BroadcastAddress != "" && net.ParseIP(rpConfig.BroadcastAddress) == nil {
+		return fmt.Errorf("ringpop config malformed `broadcastAddress` param")
+	}
+	return nil
+}
 
+// GetMembershipMonitor return a membership monitor
+func (factory *ringpopFactory) GetMembershipMonitor() (membership.Monitor, error) {
 	return factory.getMembership()
 }
 
-func (factory *RingpopFactory) getMembership() (membership.Monitor, error) {
-	if factory.membershipMonitor != nil {
-		return factory.membershipMonitor, nil
-	}
-
-	membershipMonitor, err := factory.createMembership()
-	if err != nil {
-		return nil, err
-	}
-	factory.membershipMonitor = membershipMonitor
-	return membershipMonitor, nil
+func (factory *ringpopFactory) getMembership() (membership.Monitor, error) {
+	var err error
+	factory.monOnce.Do(func() {
+		if rp, err := ringpop.New("temporal", ringpop.Channel(factory.getTChannel()), ringpop.AddressResolverFunc(factory.broadcastAddressResolver)); err != nil {
+			factory.logger.Fatal("Failed to get new ringpop", tag.Error(err))
+		} else {
+			mrp := membership.NewRingPop(rp, factory.config.MaxJoinDuration, factory.logger)
+
+			factory.membershipMonitor = membership.NewRingpopMonitor(
+				factory.serviceName,
+				factory.servicePortMap,
+				mrp,
+				factory.logger,
+				factory.metadataManager,
+				factory.broadcastAddressResolver,
+			)
+		}
+	})
+
+	return factory.membershipMonitor, err
 }
 
-func (factory *RingpopFactory) createMembership() (membership.Monitor, error) {
-	// use actual listen port (in case service is bound to :0 or 0.0.0.0:0)
-	rp, err := factory.getRingpop()
-	if err != nil {
-		return nil, fmt.Errorf("ringpop creation failed: %v", err)
-	}
-
-	membershipMonitor := membership.NewRingpopMonitor(factory.serviceName,
-		factory.servicePortMap, rp, factory.logger, factory.metadataManager, factory.broadcastAddressResolver)
+func (factory *ringpopFactory) broadcastAddressResolver() (string, error) {
+	return membership.BuildBroadcastHostPort(factory.getTChannel().PeerInfo(), factory.config.BroadcastAddress)
+}
 
-	return membershipMonitor, nil
+func (factory *ringpopFactory) getTChannel() *tchannel.Channel {
+	factory.chOnce.Do(func() {
+		ringpopServiceName := fmt.Sprintf("%v-ringpop", factory.serviceName)
+		ringpopHostAddress := net.JoinHostPort(factory.getListenIP().String(), convert.IntToString(factory.rpcConfig.MembershipPort))
+		enableTLS := factory.dc.GetBoolProperty(dynamicconfig.EnableRingpopTLS, false)()
+
+		if enableTLS {
+			if clientTLSConfig, err := factory.tlsFactory.GetInternodeClientConfig(); err != nil {
+				factory.logger.Fatal("Failed to get internode TLS client config", tag.Error(err))
+			} else {
```

Comment:
code here depends on the fact that logger.Fatal() would exit process, so we don't need to use else here, it would help reduce the indentation by a lot here.

Reply:
Yeah, I got a little over zealous with the else statements.